x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Bucket.java

```diff
@@ -89,41 +94,49 @@ public class Bucket extends EsqlScalarFunction implements Validatable {
-        @Param(name = "from", type = { "integer", "long", "double", "date", "keyword", "text" }) Expression from,
-        @Param(name = "to", type = { "integer", "long", "double", "date", "keyword", "text" }) Expression to
+        @Param(name = "bucketsOrSpan", type = { "integer", "double", "date_period", "time_duration" }) Expression bucketsOrSpan,
+        @Param(name = "from", type = { "integer", "long", "double", "date", "keyword", "text" }, optional = true) Expression from,
+        @Param(name = "to", type = { "integer", "long", "double", "date", "keyword", "text" }, optional = true) Expression to
     ) {
-        super(source, List.of(field, buckets, from, to));
+        super(source, from != null && to != null ? List.of(field, bucketsOrSpan, from, to) : List.of(field, bucketsOrSpan));
         this.field = field;
-        this.buckets = buckets;
+        this.bucketsOrSpan = bucketsOrSpan;
         this.from = from;
         this.to = to;
     }
 
     @Override
     public boolean foldable() {
-        return field.foldable() && buckets.foldable() && from.foldable() && to.foldable();
+        return field.foldable() && bucketsOrSpan.foldable() && (from == null || from.foldable()) && (to == null || to.foldable());
     }
 
     @Override
     public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
-        int b = ((Number) buckets.fold()).intValue();
-
         if (field.dataType() == DataTypes.DATETIME) {
-            long f = foldToLong(from);
-            long t = foldToLong(to);
-            return DateTrunc.evaluator(
-                source(),
-                toEvaluator.apply(field),
-                new DateRoundingPicker(b, f, t).pickRounding().prepareForUnknown()
-            );
+            Rounding.Prepared preparedRounding;
+            if (bucketsOrSpan.dataType().isInteger()) {
+                int b = ((Number) bucketsOrSpan.fold()).intValue();
+                long f = foldToLong(from);
+                long t = foldToLong(to);
+                preparedRounding = new DateRoundingPicker(b, f, t).pickRounding().prepareForUnknown();
+            } else {
+                preparedRounding = DateTrunc.createRounding(bucketsOrSpan.fold(), DEFAULT_TZ);
+            }
+            return DateTrunc.evaluator(source(), toEvaluator.apply(field), preparedRounding);
         }
         if (field.dataType().isNumeric()) {
-            double f = ((Number) from.fold()).doubleValue();
-            double t = ((Number) to.fold()).doubleValue();
+            double r;
+            if (from != null) {
+                int b = ((Number) bucketsOrSpan.fold()).intValue();
+                double f = ((Number) from.fold()).doubleValue();
+                double t = ((Number) to.fold()).doubleValue();
+                r = pickRounding(b, f, t);
+            } else {
+                r = ((Number) bucketsOrSpan.fold()).doubleValue();
```

Comment:
A `0d` will result in a `NaN`, but [this isn't new](https://github.com/elastic/elasticsearch/issues/105166) and would address it subsequently.